### ISSUE TITLE
Global asset helper for dynamic JS imports on Vapor

### ISSUE THREAD

**Issue by AtumG** [archtechx/tenancy#261]
I have an issue with global asset urls on vapor when using dynamic imports in my JS files.

My javascript files are using dynamic imports and everything is working flawlessly on my local machine. 

When I deploy to vapor I am not getting the correct asset urls for the dynamic imports only. The main `tenants.js` file loads correctly over cloudfront, but the dynamic import `CompanySelect.vue` gets a 404 because it's trying to load from the tenant domain.

In other words:
Expecting: `https://abcdefghijkl.cloudfront.net/123-456-789/js/tenant/components/forms/company-select.js`
But the app is attempting to load: `https://foo.redacted.dev/js/tenant/components/forms/company-select.js` and getting a 404 error

![Screenshot 2020-01-03 22 19 41](https://user-images.githubusercontent.com/36754831/71760740-2a118a80-2e77-11ea-9345-7f78b0a39aa6.png)

![Screenshot 2020-01-03 21 59 56](https://user-images.githubusercontent.com/36754831/71760564-68597a80-2e74-11ea-90b5-bc46253417f8.png)

Here's the context:

~~~javascript
// webpack.mix.js
const mix = require('laravel-mix');

mix.babelConfig({
    plugins: ['@babel/plugin-syntax-dynamic-import'],
  });

mix.js('resources/js/app.js', 'public/js');

mix.js('resources/js/tenant/tenant', 'public/js/tenant');

mix.sass('resources/sass/tenant/app.scss', 'public/css/tenant/app.css');
mix.sass('resources/sass/app.scss', 'public/css/app.css');

if (mix.inProduction()) {
    const ASSET_URL = process.env.ASSET_URL + "/";

    mix.webpackConfig(webpack => {
        return {
            plugins: [
                new webpack.DefinePlugin({
                    "process.env.ASSET_PATH": JSON.stringify(ASSET_URL)
                })
            ],
            output: {
                publicPath: ASSET_URL
            }
        };
    });
}
~~~

~~~html
//At the end of my layout.blade.php file
<script src="{{global_asset('js/tenant/tenant.js')}}"></script>
~~~

~~~javascript
// resources/js/tenant/tenant.js

const tenant = new Vue({
    el: '#tenant',
    components: {
        CompanySelect : () => import(/* webpackChunkName: "js/tenant/components/forms/company-select" */ '../tenant/components/forms/CompanySelect.vue'),
       }
})
~~~

To reiterate, code splitting is working great locally but components are missing once deployed to Laravel Vapor.

Thanks in advance for any support!

**Comment by stancl** [comment]
Hi,

So the issue is only with code splitting? `global_asset()` itself works correctly with Vapor?

I wonder if the `ASSET_URL` can be an issue. Vapor injects that environment variable and my package suffixes it with `tenant/$tenant_id/` for `asset()` (but probably not `global_asset()` calls.

This will probably have no effect, but maybe try using `mix()` instead of `global_asset()`.

Also what version of Mix are you using? I know that v4 is a bit weird when it comes to dynamic imports.

**Comment by AtumG** [comment]
Yes, the issue is only with code splitting. Using global_asset() does work with Vapor for the main js file but not the chunks.

I concur with your theory about global_asset() not getting the vapor prefix. I switched global_asset() to mix() with the same result.

I'm on mix 4.0.7 according to my package.json file.

**Comment by stancl** [comment]
IIRC Mix 4 uses Webpack 4 which has some weird issues with code splitting. Mix 5 was released in September, but I'm not sure if it uses Webpack 5 because it's still in beta.

You could also try using Mix 3. Webpack 3 doesn't have issues with code splitting.

Also, you have this in your Mix config:

```js
output: {
    publicPath: ASSET_URL
}
```

Looks correct, there's also a `chunkFilename` config but that should be respecting `publicPath` afaik. It might be worth inspecting the `ASSET_URL`, but I have no clue how you'd do that.

Vapor should be injecting the environment variable, but it doesn't seem injected. The JS file loads correctly because the helpers work correctly, but Mix uses wrong paths for code splitting. Maybe also make sure Vapor is building your assets, instead of you just pushing locally built assets.

**Comment by AtumG** [comment]
After digging further - it looks like the tenancy package is doing its job correctly. The issue lies within webpack/laravel-mix.

I came up with a working solution that seems pretty straightforward.

I upgraded to Mix 5 and I'm using `global_asset()`. All the webpack chunks are loaded properly and work in my local environment. 

Things go wrong when uploaded to Vapor. Vapor correctly injects it's asset url at build time so the manifest file is accurate (includes vapor deployment id). Once the js file is loaded, since it was loaded on the main domain, webpack assumes this is the location of any chunk files. (i.e `https://tenant.mysite.com/js/wherearemyfiles.js` instead of `https://hash.cloudfront.net/vapor-deploy-id/js/heretheyare.js`)

To fix the problem, I had to set the `__webpack_public_path__` at runtime and parse the location the original js file was loaded from in order to get the base path for the chunks. 

In other words, simply add the following lines to the top of `resources/js/bootstrap.js`

~~~javascript
const url = new URL(document.currentScript.src);
__webpack_public_path__ = (url.href.substr(0, url.href.indexOf('/js')) + __webpack_public_path__);
~~~

After adding these 2 lines and deploying to Vapor, my chunks now load properly from cloudfront.

I'm still getting some issues with dynamic imports in both local and on vapor but I don't think that is related to this thread. I'll try all this with Mix 3 and see if that fixes things.

**Comment by AtumG** [comment]
Thank you for pointing me in the right direction with the problem being js related rather than the tenancy package!

After fixing import conflicts and using Mix 5 with the public path hack above, I've got things working correctly on Vapor. I tried to downgrade to Mix 3 but had more issues than just using Mix 5... at least with Chrome. 

I feel like there has to be a more elegant way to handle this but I'm out of ideas and moving on with this workable solution for now. Thanks again for the timely support!
